SDL_SetWindowFullscreen fails with 'Couldn't find any matching video modes' when using dummy driver

This is one of the last issues I'm experiencing with the testsuite from py-sdl/py-sdl2 in Fedora after switching to sdl2-compat[^thanks].

I'm using sdl2-compat 2.32.54 with py-sdl/py-sdl2/pull/280 applied as a patch on top. Running the testsuite with the dummy video driver (SDL_VIDEODRIVER=dummy) [`test_SDL_SetWindowFullscreen`](https://github.com/py-sdl/py-sdl2/blob/6414ee1c5f4a6eb91b71f5f9e35d469eee395b9f/sdl2/test/video_test.py#L640C1-L655C39) fails.

I modified the test a little in order to get the actual error message:

```diff
--- a/sdl2/test/video_test.py
+++ b/sdl2/test/video_test.py
@@ -646,10 +646,10 @@ def test_SDL_SetWindowFullscreen(with_sdl):
     is_fullscreen = sdl2.SDL_WINDOW_FULLSCREEN
     for flag in flags:
         window = sdl2.SDL_CreateWindow(b"Test", 0, 0, 1024, 768, flag)
-        sdl2.SDL_SetWindowFullscreen(window, True)
+        assert sdl2.SDL_SetWindowFullscreen(window, True) == 0, _check_error_msg()
         flags = sdl2.SDL_GetWindowFlags(window)
         assert flags & is_fullscreen == is_fullscreen
-        sdl2.SDL_SetWindowFullscreen(window, False)
+        assert sdl2.SDL_SetWindowFullscreen(window, False) == 0, _check_error_msg()
         flags = sdl2.SDL_GetWindowFlags(window)
         assert flags & is_fullscreen != is_fullscreen
         sdl2.SDL_DestroyWindow(window)
```

With that the test reports: "AssertionError: Couldn't find any matching video modes". I also tried the same test outside our build environment on X11 and Wayland. Both work.

With SDL2 2.30.11 the test also passes with the dummy driver.

[^thanks]: Thanks for fixing the majority of issues reported downstream.

--- a/src/sdl2_compat.c
+++ b/src/sdl2_compat.c
@@ -8209,6 +8209,14 @@ DisplayMode_3to2(const SDL_DisplayMode *in, SDL2_DisplayMode *out) {
         out->h = in->h;
         out->refresh_rate = (int) SDL3_lroundf(in->refresh_rate);
         out->driverdata = in->internal;
+
+        /* Make sure the returned refresh rate and format are something valid */
+        if (!out->refresh_rate) {
+            out->refresh_rate = 60;
+        }
+        if (!out->format) {
+            out->format = SDL_PIXELFORMAT_XRGB8888;
+        }
     }
 }
 
@@ -8495,11 +8503,6 @@ SDL_GetWindowDisplayMode(SDL_Window *window, SDL2_DisplayMode *mode)
         if (SDL_GetDesktopDisplayMode(display, mode) < 0) {
             return -1;
         }
-
-        /* When returning the desktop mode, make sure the refresh is some nonzero value. */
-        if (mode->refresh_rate == 0) {
-            mode->refresh_rate = 60;
-        }
     } else {
         if (!SDL_GetClosestDisplayMode(display, mode, mode)) {
             SDL_zerop(mode);
@@ -8533,14 +8536,6 @@ SDL_GetClosestDisplayMode(int displayIndex, const SDL2_DisplayMode *mode, SDL2_D
     }
 
     DisplayMode_3to2(ret3, closest);
-
-    /* Make sure the returned refresh rate and format are something valid */
-    if (!closest->refresh_rate) {
-        closest->refresh_rate = mode->refresh_rate ? mode->refresh_rate : 60;
-    }
-    if (!closest->format) {
-        closest->format = SDL_PIXELFORMAT_XRGB8888;
-    }
     return closest;
 }
 

--- a/src/sdl2_compat.c
+++ b/src/sdl2_compat.c
@@ -8454,6 +8454,14 @@ static int ApplyFullscreenMode(SDL_Window *window)
         mode.h = property->h;
         mode.refresh_rate = (float)property->refresh_rate;
     } else {
+        int count = 0;
+        SDL3_free(SDL3_GetFullscreenDisplayModes(displayID, &count));
+
+        /* If the video driver has no fullscreen modes, transition to fullscreen desktop mode */
+        if (count == 0) {
+            return 0;
+        }
+
         SDL3_GetWindowSize(window, &mode.w, &mode.h);
     }
 
@@ -8463,6 +8471,15 @@ static int ApplyFullscreenMode(SDL_Window *window)
     if (mode.displayID && SDL3_SetWindowFullscreenMode(window, &mode)) {
         return 0;
     }
+    else if (property) {
+        const SDL_DisplayMode *desktop = SDL3_GetDesktopDisplayMode(displayID);
+
+        /* It's possible that this was an attempted transition to the desktop display mode on a driver without modesetting support */
+        if (desktop && desktop->w == property->w && desktop->h == property->h) {
+            return 0;
+        }
+    }
+
     return -1;
 }
 

